### PR TITLE
uf-8546: Fetching neighborhood notifications.

### PR DESCRIPTION
**Pull request by mha109**

### CODE REVIEW

File: src/main/java/se/sundsvall/byggrintegrator/api/validation/ValidPersonalOrOrgNumber.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+@Documented
+@ConstraintComposition(CompositionType.OR)
+@Constraint(validatedBy = {})
+@ValidPersonalNumber
+@ValidOrganizationNumber
+@ReportAsSingleViolation
+@Target({ ElementType.FIELD, ElementType.PARAMETER })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ValidPersonalOrOrgNumber {
```
Comment (jklingst) [inline src/main/java/se/sundsvall/byggrintegrator/api/validation/ValidPersonalOrOrgNumber.java:27]: This works, but you could also re-use existing `@ValidPersonalNumber` and `@ValidOrganizationNumber` with `org.hibernate.validator.constraints.ConstraintComposition`. See org.hibernate.validator.constraints.ConstraintComposition

File: src/main/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapper.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	public List<NeighborhoodNotificationsDto> mapToNeighborhoodNotificationsDto(GetRelateradeArendenByPersOrgNrAndRoleResponse response) {
+		var errands = Optional.ofNullable(response.getGetRelateradeArendenByPersOrgNrAndRoleResult())
+			.flatMap(result -> Optional.ofNullable(result.getArende())).stream()
+				.flatMap(Collection::stream)
+			.toList();
+
+		List<NeighborhoodNotificationsDto> neighborhoodNotificationsDtoList = new ArrayList<>();
+
+		// Collect the info we want from errands that have a valid event
+		errands.forEach(arende -> {
```
Comment (jklingst) [inline src/main/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapper.java:61]: return `errands.stream().filter(...).map(...)`?

File: src/test/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	void testMapToNeighborhoodNotificationsDto() {
+		// Arrange
+		var response = createRelateradeArendenResponse();
+
+		// Act
+		var neighborNotificationsDtos = mapper.mapToNeighborhoodNotificationsDto(response);
+
+		// Assert
+		assertThat(neighborNotificationsDtos).hasSize(1);
+		assertThat(neighborNotificationsDtos.getFirst().getByggrErrandNumber()).isEqualTo(BYGGR_ARENDE_NR_1);
```
Comment (jklingst) [inline src/test/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapperTest.java:40]: "Wrap" everything beneath `neighborNotificationsDtos.getFirst()` in a `.satisifies(...)` block ?

File: src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import se.sundsvall.byggrintegrator.api.model.KeyValue;
+import se.sundsvall.byggrintegrator.model.NeighborhoodNotificationsDto;
+
+@Component
+public class ApiResponseMapper {
+
+	public List<KeyValue> mapToKeyValueResponseList(List<NeighborhoodNotificationsDto> errands) {
+		List<KeyValue> response = new ArrayList<>();
+
+		errands.forEach(
```
Comment (jklingst) [inline src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java:17]: Null-check on the `errands` parameter ?

File: src/test/java/se/sundsvall/byggrintegrator/api/ByggrIntegratorResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+@ActiveProfiles("junit")
+@SpringBootTest(classes = Application.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class ByggrIntegratorResourceTest {
+
+	@MockBean
+	private ByggrIntegratorService mockByggrIntegratorService;
+
+	@Autowired
+	private ByggrIntegratorResource resource;
```
Comment (josc04) [inline src/test/java/se/sundsvall/byggrintegrator/api/ByggrIntegratorResourceTest.java:39]: Not used?

File: src/test/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		assertThat(neighborNotificationsDtos.getFirst().getPropertyDesignation()).hasSize(2);
+		assertThat(neighborNotificationsDtos.getFirst().getPropertyDesignation().getFirst()).satisfies(propertyDesignation -> {
+			assertThat(propertyDesignation.getProperty()).isEqualTo("ANKEBORG");
+			assertThat(propertyDesignation.getDesignation()).isEqualTo("1:1234");
+		});
+		assertThat(neighborNotificationsDtos.getFirst().getPropertyDesignation().getLast()).satisfies(propertyDesignation -> {
+			assertThat(propertyDesignation.getProperty()).isEqualTo("ANKEBORG");
+			assertThat(propertyDesignation.getDesignation()).isEqualTo("2:5678");
+		});
```
Comment (josc04) [inline src/test/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapperTest.java:49]: Warning for OCD. This is another way to write the test:
```
		assertThat(neighborNotificationsDtos.getFirst().getPropertyDesignation()).satisfiesExactlyInAnyOrder(propertyDesignation -> {
			assertThat(propertyDesignation.getProperty()).isEqualTo("ANKEBORG");
			assertThat(propertyDesignation.getDesignation()).isEqualTo("1:1234");
		}, propertyDesignation -> {
			assertThat(propertyDesignation.getProperty()).isEqualTo("ANKEBORG");
			assertThat(propertyDesignation.getDesignation()).isEqualTo("2:5678");
		});
```

File: src/test/java/se/sundsvall/byggrintegrator/service/ApiResponseMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		// Arrange
+		var errands = generateNeighborNotificationsDtos();
+
+		// Act
+		var keyValues = apiResponseMapper.mapToKeyValueResponseList(errands);
+
+		// Assert
+		assertThat(keyValues).hasSize(4);
+		assertThat(keyValues).extracting(KeyValue::key).containsExactly("dnr123", "dnr123", "dnr456", "dnr456");
+		assertThat(keyValues).extracting(KeyValue::value).containsExactly("dnr123, des-1 type1", "dnr123, des-2 type2", "dnr456, des-3 type3", "dnr456, des-4 type4");
```
Comment (josc04) [inline src/test/java/se/sundsvall/byggrintegrator/service/ApiResponseMapperTest.java:27]: Change to containsExactlyInAnyOrder to account for differences in different execution environments?

File: src/test/java/se/sundsvall/byggrintegrator/service/ByggrIntegratorServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		assertThat(neighborNotifications).hasSize(2);
+		assertThat(neighborNotifications.getFirst().key()).isEqualTo("BYGG 2024-000123");
+		assertThat(neighborNotifications.getFirst().value()).isEqualTo("BYGG 2024-000123, ANKEBORG 1:1234");
+		assertThat(neighborNotifications.getLast().key()).isEqualTo("BYGG 2024-000123");
+		assertThat(neighborNotifications.getLast().value()).isEqualTo("BYGG 2024-000123, ANKEBORG 2:5678");
+		verify(mockByggrIntegration).getRoles();
+		verify(mockByggrIntegration).getErrandsFromByggr(IDENTIFIER, ROLES);
+		verify(mockByggrIntegrationMapper).mapToNeighborhoodNotificationsDto(any(GetRelateradeArendenByPersOrgNrAndRoleResponse.class));
+		verify(mockApiResponseMapper).mapToKeyValueResponseList(anyList());
+		verifyNoMoreInteractions(mockByggrIntegration, mockByggrIntegrationMapper);
```
Comment (josc04) [inline src/test/java/se/sundsvall/byggrintegrator/service/ByggrIntegratorServiceTest.java:70]: Add `mockApiResponseMapper` in verifyNoMoreInteractions (comment is applicable for all places where verifyNoMoreInteractions is called)?

File: src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java (hunk 4 above)
Comment (josc04) [inline src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java:17]: Cover for possible NPE: `Optional.ofNullable(errands).orElse(Collections.emptyList()).forEach(`?

File: src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java
Comment (josc04) [inline src/main/java/se/sundsvall/byggrintegrator/service/ApiResponseMapper.java:1]: Comment marked as resolved, but no changes has been made..?

Suggestion to change to:
```
@Component
public class ApiResponseMapper {

	public List<KeyValue> mapToKeyValueResponseList(List<NeighborhoodNotificationsDto> errands) {
		return Optional.ofNullable(errands).orElse(Collections.emptyList()).stream()
			.map(this::mapDesignations)
			.flatMap(List::stream)
			.toList();
	}

	private List<KeyValue> mapDesignations(NeighborhoodNotificationsDto errand) {
		return Optional.ofNullable(errand.getPropertyDesignation()).orElse(Collections.emptyList()).stream()
			.map(designation -> mapToKeyValue(errand.getByggrErrandNumber(), designation))
			.toList();
	}

	private KeyValue mapToKeyValue(String dnr, NeighborhoodNotificationsDto.PropertyDesignation designation) {
		return new KeyValue(dnr, dnr + ", " + designation.getProperty() + " " + designation.getDesignation());
	}
}
```

File: src/main/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapper.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+			if(hasValidHandelseList(arende.getHandelseLista().getHandelse())) {
+				neighborhoodNotificationsDtoList.add(NeighborhoodNotificationsDto.builder()
+					.withByggrErrandNumber(arende.getDnr())
+					.withPropertyDesignation(mapToPropertyDesignations(arende.getObjektLista().getAbstractArendeObjekt()))
```
Comment (josc04) [inline src/main/java/se/sundsvall/byggrintegrator/integration/byggr/ByggrIntegrationMapper.java:65]: If you want to use the stream-variant, this is an alternative solution:
```
		// Collect the info we want from errands that have a valid event
		return errands.stream()
			.filter(arende -> hasValidHandelseList(arende.getHandelseLista().getHandelse()))
			.map(arende -> NeighborhoodNotificationsDto.builder()
				.withByggrErrandNumber(arende.getDnr())
				.withPropertyDesignation(mapToPropertyDesignations(arende.getObjektLista().getAbstractArendeObjekt()))
				.build())
			.toList();
```

**Review by josc04 — APPROVED** [review]

**Pull request merged**